### ISSUE TITLE
saltutil.refresh_pillar broke connection betweens minions and master

### ISSUE THREAD

**Issue by calve** [saltstack/salt#22503]
Hello all.

I experienced the following bug : 

1. Runned ``sudo salt '*' saltutil.refresh_pillar`` on the master
2. Every minion responded ``True``
3. Runned ``sudo salt-run manage.status``
4. All my minions are now down (except the minion running on the same host as the master)

The logs files located at ``minions://var/log/salt/minion`` contains no evidence of an unexpected behaviour nor exception.
All hosts are running Ubuntu 14.04 or 12.04. Also, on minions, it looks like the ``salt-minion`` service is still running :

    $ sudo service salt-minion status
    salt-minion start/running, process 7452

So now, I have a total of 0 remote minions responding to their master calls, and I have the pain to ssh on each of them to restart the ``salt-minion`` service.

I think, but cannot proove it, that this issue may be related to issuer #13558 as I experience the same symptoms. I experienced this bug once at the moment, but I will provides extra informations if it append again.

Please, let me know if you need more informations to inspect this behaviour

**Comment by cachedout** [comment]
If this is with multi-master, it is a duplicate of #19708

**Comment by rallytime** [comment]
@calve Can you tell us what version of salt you're running? Please post the output of `salt --versions-report` for the master and `salt <minion> test.versions_report` for one of the affected minions.

Also, as @cachedout mentioned, this is a known bug with multi-master. Are you using a multi-master setup?

**Comment by calve** [comment]
I am not using a multple master setup. Every minions refer to the master as ``10.x.y.z`` The minion running on the same host as the master however refers to the master as ``localhost``

Versions report : 
 master:
```
 salt --versions-report                             
           Salt: 2014.7.0
         Python: 2.7.6 (default, Mar 22 2014, 22:59:56)
         Jinja2: 2.7.2
       M2Crypto: 0.21.1
 msgpack-python: 0.3.0
   msgpack-pure: Not Installed
       pycrypto: 2.6.1
        libnacl: Not Installed
         PyYAML: 3.10
          ioflo: Not Installed
          PyZMQ: 14.0.1
           RAET: Not Installed
            ZMQ: 4.0.4
           Mako: 0.9.1
```

minions, not all, but a representative sample ( bespin runs ubuntu 12.04, the two other are under 14.04) :
```
bespin:
               Salt: 2014.7.0
             Python: 2.7.3 (default, Feb 27 2014, 19:58:35)
             Jinja2: 2.6
           M2Crypto: 0.21.1
     msgpack-python: 0.1.10
       msgpack-pure: Not Installed
           pycrypto: 2.4.1
            libnacl: Not Installed
             PyYAML: 3.10
              ioflo: Not Installed
              PyZMQ: 13.0.0
               RAET: Not Installed
                ZMQ: 3.2.2
               Mako: Not Installed
dagobah:
               Salt: 2014.7.0
             Python: 2.7.6 (default, Mar 22 2014, 22:59:56)
             Jinja2: 2.7.2
           M2Crypto: 0.21.1
     msgpack-python: 0.3.0
       msgpack-pure: Not Installed
           pycrypto: 2.6.1
            libnacl: Not Installed
             PyYAML: 3.10
              ioflo: Not Installed
              PyZMQ: 14.0.1
               RAET: Not Installed
                ZMQ: 4.0.4
               Mako: 0.9.1
abregadorae:
               Salt: 2014.7.0
             Python: 2.7.6 (default, Mar 22 2014, 22:59:56)
             Jinja2: 2.7.2
           M2Crypto: 0.21.1
     msgpack-python: 0.3.0
       msgpack-pure: Not Installed
           pycrypto: 2.6.1
            libnacl: Not Installed
             PyYAML: 3.10
              ioflo: Not Installed
              PyZMQ: 14.0.1
               RAET: Not Installed
                ZMQ: 4.0.4
               Mako: 0.9.1

```

Thanks for your interest

**Comment by teizz** [comment]
I'm getting this after a `grains.append roles newrole`, which also does a pillar refresh at some point. The salt-minion returns to a functioning state after a `killall -SIGUSR1 salt-minion`. After this SIGUSR1 it also shows the salt-minion returning my test.ping commands which had timed-out on the salt-master.

Below also a versions-report.

```
[INFO    ] User mattijs Executing command grains.append with jid 20150416150510210282
[DEBUG   ] Command details {'tgt_type': 'glob', 'jid': '20150416150510210282', 'tgt': 'salt-minion', 'ret': '', 'user': 'mattijs', 'arg': ['roles', 'newrole'], 'fun': 'grains.append'}
[INFO    ] Starting a new job with PID 5521
[DEBUG   ] Decrypting the current master AES key
[DEBUG   ] Loaded minion key: /etc/salt/pki/minion/minion.pem
[DEBUG   ] Reading configuration from /etc/salt/minion
[DEBUG   ] Decrypting the current master AES key
[DEBUG   ] Loaded minion key: /etc/salt/pki/minion/minion.pem
[DEBUG   ] Loaded minion key: /etc/salt/pki/minion/minion.pem
[INFO    ] Loading fresh modules for state activity
[DEBUG   ] Fetching file from saltenv 'authentication', ** attempting ** 'salt://top.sls'
[DEBUG   ] Fetching file from saltenv 'interfaces', ** attempting ** 'salt://top.sls'
[DEBUG   ] Fetching file from saltenv 'base', ** attempting ** 'salt://top.sls'
[INFO    ] Creating module dir '/var/cache/salt/minion/extmods/grains'
[INFO    ] Syncing grains for environment 'base'
[INFO    ] Loading cache from salt://_grains, for base)
[DEBUG   ] Decrypting the current master AES key
[DEBUG   ] Loaded minion key: /etc/salt/pki/minion/minion.pem
[INFO    ] Caching directory '_grains/' for environment 'base'
[DEBUG   ] Local cache dir: '/var/cache/salt/minion/files/base/_grains'
[DEBUG   ] SaltEvent PUB socket URI: ipc:///var/run/salt/minion/minion_event_f5b35bdb6a_pub.ipc
[DEBUG   ] SaltEvent PULL socket URI: ipc:///var/run/salt/minion/minion_event_f5b35bdb6a_pull.ipc
[DEBUG   ] Sending event - data = {'_stamp': '2015-04-16T13:05:13.079081'}
[DEBUG   ] Handling event 'module_refresh||||||\x81\xa6_stamp\xba2015-04-16T13:05:13.079081'
[DEBUG   ] Reading configuration from /etc/salt/minion
[DEBUG   ] SaltEvent PUB socket URI: ipc:///var/run/salt/minion/minion_event_f5b35bdb6a_pub.ipc
[DEBUG   ] SaltEvent PULL socket URI: ipc:///var/run/salt/minion/minion_event_f5b35bdb6a_pull.ipc
[DEBUG   ] Sending event - data = {'_stamp': '2015-04-16T13:05:13.106552'}
[INFO    ] Returning information for job: 20150416150510210282
[DEBUG   ] Decrypting the current master AES key
[DEBUG   ] Loaded minion key: /etc/salt/pki/minion/minion.pem
[DEBUG   ] Handling event 'pillar_refresh||||||\x81\xa6_stamp\xba2015-04-16T13:05:13.106552'
[DEBUG   ] Decrypting the current master AES key
[DEBUG   ] Loaded minion key: /etc/salt/pki/minion/minion.pem
======== Salt Debug Stack Trace =========
  File "/usr/bin/salt-minion", line 14, in <module>
    salt_minion()
  File "/usr/lib/python2.7/dist-packages/salt/scripts.py", line 57, in salt_minion
    minion.start()
  File "/usr/lib/python2.7/dist-packages/salt/__init__.py", line 264, in start
    self.minion.tune_in()
  File "/usr/lib/python2.7/dist-packages/salt/minion.py", line 1611, in tune_in
    self.pillar_refresh()
  File "/usr/lib/python2.7/dist-packages/salt/minion.py", line 1416, in pillar_refresh
    self.opts['environment'],
  File "/usr/lib/python2.7/dist-packages/salt/pillar/__init__.py", line 91, in compile_pillar
    ret_pillar = self.sreq.crypted_transfer_decode_dictentry(load, dictkey='pillar', tries=3, timeout=7200)
  File "/usr/lib/python2.7/dist-packages/salt/transport/__init__.py", line 243, in crypted_transfer_decode_dictentry
    ret = self.sreq.send('aes', self.auth.crypticle.dumps(load), tries, timeout)
  File "/usr/lib/python2.7/dist-packages/salt/payload.py", line 214, in send
    polled = self.poller.poll(timeout * 1000)
  File "/usr/lib/python2.7/dist-packages/zmq/sugar/poll.py", line 110, in poll
    return zmq_poll(self.sockets, timeout=timeout)
=========================================
[DEBUG   ] Exception while handling events
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/minion.py", line 1611, in tune_in
    self.pillar_refresh()
  File "/usr/lib/python2.7/dist-packages/salt/minion.py", line 1416, in pillar_refresh
    self.opts['environment'],
  File "/usr/lib/python2.7/dist-packages/salt/pillar/__init__.py", line 91, in compile_pillar
    ret_pillar = self.sreq.crypted_transfer_decode_dictentry(load, dictkey='pillar', tries=3, timeout=7200)
  File "/usr/lib/python2.7/dist-packages/salt/transport/__init__.py", line 243, in crypted_transfer_decode_dictentry
    ret = self.sreq.send('aes', self.auth.crypticle.dumps(load), tries, timeout)
  File "/usr/lib/python2.7/dist-packages/salt/payload.py", line 214, in send
    polled = self.poller.poll(timeout * 1000)
  File "/usr/lib/python2.7/dist-packages/zmq/sugar/poll.py", line 110, in poll
    return zmq_poll(self.sockets, timeout=timeout)
  File "_poll.pyx", line 115, in zmq.backend.cython._poll.zmq_poll (zmq/backend/cython/_poll.c:1677)
  File "checkrc.pxd", line 21, in zmq.backend.cython.checkrc._check_rc (zmq/backend/cython/_poll.c:2047)
ZMQError: Interrupted system call
[INFO    ] User mattijs Executing command test.ping with jid 20150416150543179746
[DEBUG   ] Command details {'tgt_type': 'glob', 'jid': '20150416150543179746', 'tgt': 'salt-minion', 'ret': '', 'user': 'mattijs', 'arg': [], 'fun': 'test.ping'}
[INFO    ] User mattijs Executing command saltutil.find_job with jid 20150416150603189133
[INFO    ] Starting a new job with PID 5611
[DEBUG   ] Command details {'tgt_type': 'glob', 'jid': '20150416150603189133', 'tgt': 'salt-minion', 'ret': '', 'user': 'mattijs', 'arg': ['20150416150543179746'], 'fun': 'saltutil.find_job'}
[INFO    ] Returning information for job: 20150416150543179746
[INFO    ] User mattijs Executing command test.ping with jid 20150416150646777592
[DEBUG   ] Command details {'tgt_type': 'glob', 'jid': '20150416150646777592', 'tgt': 'salt-minion', 'ret': '', 'user': 'mattijs', 'arg': [], 'fun': 'test.ping'}
[INFO    ] Starting a new job with PID 5616
[INFO    ] Returning information for job: 20150416150603189133
[INFO    ] User mattijs Executing command saltutil.find_job with jid 20150416150706794967
[DEBUG   ] Command details {'tgt_type': 'glob', 'jid': '20150416150706794967', 'tgt': 'salt-minion', 'ret': '', 'user': 'mattijs', 'arg': ['20150416150646777592'], 'fun': 'saltutil.find_job'}
[INFO    ] Starting a new job with PID 5621
[INFO    ] Returning information for job: 20150416150646777592
[INFO    ] Starting a new job with PID 5626
[INFO    ] Returning information for job: 20150416150706794967
[DEBUG   ] Decrypting the current master AES key
[DEBUG   ] Loaded minion key: /etc/salt/pki/minion/minion.pem
[DEBUG   ] Decrypting the current master AES key
[DEBUG   ] Loaded minion key: /etc/salt/pki/minion/minion.pem
[DEBUG   ] Decrypting the current master AES key
[DEBUG   ] Loaded minion key: /etc/salt/pki/minion/minion.pem
[DEBUG   ] Decrypting the current master AES key
[DEBUG   ] Loaded minion key: /etc/salt/pki/minion/minion.pem
[INFO    ] User mattijs Executing command test.ping with jid 20150416151041560883
[DEBUG   ] Command details {'tgt_type': 'glob', 'jid': '20150416151041560883', 'tgt': 'salt-minion', 'ret': '', 'user': 'mattijs', 'arg': [], 'fun': 'test.ping'}
[INFO    ] Starting a new job with PID 5639
[INFO    ] Returning information for job: 20150416151041560883
[DEBUG   ] Decrypting the current master AES key
[DEBUG   ] Loaded minion key: /etc/salt/pki/minion/minion.pem
```

```
root@salt-minion:~# salt-minion --versions-report
                  Salt: 2014.7.2
                Python: 2.7.6 (default, Mar 22 2014, 22:59:56)
                Jinja2: 2.7.2
              M2Crypto: 0.21.1
        msgpack-python: 0.3.0
          msgpack-pure: Not Installed
              pycrypto: 2.6.1
               libnacl: Not Installed
                PyYAML: 3.10
                 ioflo: Not Installed
                 PyZMQ: 14.0.1
                  RAET: Not Installed
                   ZMQ: 4.0.4
                  Mako: 0.9.1
 Debian source package: 2014.7.2+ds-1trusty2

```

**Comment by teizz** [comment]
This is fixed for me with this merge: https://github.com/saltstack/salt/pull/21465

**Comment by rallytime** [comment]
Thanks for this extra info. Apologies for not making it back to this issue sooner. @calve Can you give the pull request mentioned by @teizz a try and see if this resolves the issue for you? If not, feel free to give 2015.2.0 a try once it is released.

**Comment by teizz** [comment]
No worries :) After seeing this got fixed in the 2015.2 branch I noticed it mentioned in the 2014.7.5 release notes as well. Since I was running 2014.7.5 but it wasn't setting SO_KEEPALIVE for the ZMQ socket, I double checked the pull request. @cachedout was kind enough to merge and so 2014.7 branch is working for me too now: https://github.com/saltstack/salt/pull/23258

**Comment by cachedout** [comment]
Looks like it was just a merge conflict that went awry. Many thanks to @teizz for chasing this one down!

**Comment by rallytime** [comment]
Thanks for tracking that down @teizz! Looks like this one should be fixed in both the 2014.7 and 2015.2 branches. If we do another release on the 2014.7 branch (2014.7.6) then the fix will be there. Otherwise, the fix will be available in 2015.2.0. As such, I'm going to close this issue. However, if this pops up again, leave a comment and we will be happy to take another look. Thanks again!
